Source/WebCore/platform/graphics/Damage.h

```diff
@@ -263,11 +263,35 @@ class Damage {
     }
 
 private:
-    void initialize()
+    IntSize gridSize(int maxRectangles) const
+    {
+        const float widthToHeightRatio = static_cast<float>(m_rect.width()) / static_cast<float>(m_rect.height());
+        if (widthToHeightRatio >= 1) {
+            int gridHeight = std::max<int>(1, std::floor(std::sqrt(static_cast<float>(maxRectangles) / widthToHeightRatio)));
+            while (gridHeight > 1 && maxRectangles % gridHeight)
+                gridHeight--;
+            return { maxRectangles / gridHeight, gridHeight };
+        }
+
+        int gridWidth = std::max<int>(1, std::floor(std::sqrt(static_cast<float>(maxRectangles) * widthToHeightRatio)));
+        while (gridWidth > 1 && maxRectangles % gridWidth)
+            gridWidth--;
+        return { gridWidth, maxRectangles / gridWidth };
+    }
+
+    void initialize(uint32_t maxRectangles)
     {
         switch (m_mode) {
         case Mode::Rectangles:
-            m_gridCells = IntSize(std::ceil(static_cast<float>(m_rect.width()) / s_cellSize), std::ceil(static_cast<float>(m_rect.height()) / s_cellSize)).expandedTo({ 1, 1 });
+            if (maxRectangles) {
+                m_gridCells = gridSize(maxRectangles);
+                m_cellSize = IntSize(std::ceil(static_cast<float>(m_rect.width()) / m_gridCells.width()), std::ceil(static_cast<float>(m_rect.height()) / m_gridCells.height()));
```

Comment:
The expression we use to calculate `m_cellSize` or `m_gridCells` below is the same: an IntSize out of the ceil of operands.

There will be a constexpr ceil from C++-23, so we could extract this as a helper, `IntSize toCeiledSize(const IntRect&, const IntSize&)`. This could be inliend for now, and eventually constexpr.

Reply:
There's already ceiledIntSize, but we can't use that because it uses LayoutUnit that uses int as its value. Maybe we can just add ceiledIntSize that receives and IntSize, but not in this patch.